cve_bin_tool/sbom_manager/__init__.py

```diff
@@ -179,19 +179,29 @@ def get_vendor(self, product: str) -> list:
             vendorlist.append("UNKNOWN")
         return vendorlist
 
-    def is_valid_purl(self, purl_string: str):
+    def is_valid_string(self, string_type: str, ref_string: str) -> bool:
         """
-        Validate the PURL string is the correct form.
+        Validate the PURL, CPE string is the correct form.
 
         Args:
-        - purl_string (str): Package URL string
+        - ref_string (str): PURL, CPE strings
+        - string_type (str): ref_string type. (purl, cpe22 or cpe23)
 
         Returns:
-        - bool: True if the purl_string parameter is a valid purl string, False otherwise.
+        - bool: True if the ref_string parameter is a valid purl or cpe string, False otherwise.
 
         """
-        purl_pattern = r"^(?P<scheme>.+):(?P<type>.+)/(?P<namespace>.+)/(?P<name>.+)@(?P<version>.+)\??(?P<qualifiers>.*)#?(?P<subpath>.*)$"
-        return re.match(purl_pattern, purl_string) is not None
+        string_pattern: str
+        if string_type == "purl":
+            string_pattern = r"^(?P<scheme>.+):(?P<type>.+)/(?P<namespace>.+)/(?P<name>.+)@(?P<version>.+)\??(?P<qualifiers>.*)#?(?P<subpath>.*)$"
+
+        elif string_type == "cpe23":
+            string_pattern = r"^cpe:2\.3:[aho\*\-](:(((\?*|\*?)([a-zA-Z0-9\-\._]|(\\[\\\*\?\!\"#\$%&'\(\)\+,\-\.\/:;<=>@\[\]\^`\{\|}~]))+(\?*|\*?))|[\*\-])){5}(:(([a-zA-Z]{2,3}(-([a-zA-Z]{2}|[0-9]{3}))?)|[\*\-]))(:(((\?*|\*?)([a-zA-Z0-9\-\._]|(\\[\\\*\?\!\"#\$%&'\(\)\+,\-\.\/:;<=>@\[\]\^`\{\|}~]))+(\?*|\*?))|[\*\-])){4}"
```

Comment:
I love a ridiculous regex, but I think this is going to be hard to maintain *and* it's almost certainly going to be vulnerable to regex denial of service.

Options:
1. does anyone have a library for validating CPE strings we could leverage?
2. In lieu of this, can we split on `:` and evaluate each piece separately with an explanation of what we're looking for in the comments?
3. In lieu of this, can we split on `:` and only evaluate the parts we're going to use later?  (e.g.  `vendor, product, version = cpe[2], cpe[3], cpe[4]`)
4. In lieu of this, maybe we should validate and sanitize only `vendor, product, version` after the existing split?

I'm leaning towards the last one as potentially the right solution since that would allow us to have some util functions for `sanitize_vendor()`, `sanitize_product()`, `sanitize_version(`) that we could re-use elsewhere in purl and triage work.